### ISSUE TITLE
DataParalell copies the model onto GPUs sequentially

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#51385]
I have 8 GPUs and can see it graphically with `watch -n 0.1 nvidia-smi`. It could save some time by doing the parallel async copy?

**Comment by ezyang** [comment]
This sounds pretty reasonable

**Comment by vadimkantorov** [comment]
Maybe even scatter_all somehow could be used (e.g. we don't have bandwidth to copy to all devices at once, once the first device has received the copy, it could also transfer the model copy to other devices, halving the copy time)

**Comment by ngimel** [comment]
Are you talking about initial parameter copy? Because during training broadcast/reduce should not copy data to the cpu.

**Comment by vadimkantorov** [comment]
Yes, initial parameter copy. But the model is already on the first GPU, since I'm doing:
```
model = Model().cuda()
model = nn.DataParallel(model)
```
So it should already be amenable to inter-GPU communication (if it's enabled)

**Comment by ngimel** [comment]
We will accept a PR implementing that.
